### PR TITLE
Fix #1459: Replace UIWebView user agent creation with manual UA builder.

### LINKED ISSUE #1459: UIWebView deprecation notice
UIWebView is deprecated and may result in rejecting the app in App Store.
We have to remove it.
Currently we use it in UserAgent creation and KIF framework uses it.

### PR DESCRIPTION
**Pull request by iccub**
## Summary of Changes

The logic is based on Chromium's implementation https://chromium.googlesource.com/chromium/src.git/+/master/ios/web/common/user_agent.mm

This pull request fixes issue #1459 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

Verify new and old app versions have the same user agent.

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/internal/issues/new?assignees=&labels=security%2C+security-reviews&template=security-review.md&title=iOS:%20Security+Review+for) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Cartfile [hunk 1]
```diff
@@ -3,7 +3,6 @@ github "garvankeeley/Deferred"		"9.9.2"
 github "SnapKit/SnapKit"		"5.0.1"
 github "rs/SDWebImage"			"4.4.6"
 github "swisspol/GCDWebServer"		"3.5.2"
-github "kif-framework/KIF"		"v3.7.8"
```
Comment (iccub) [inline Cartfile:6]: We don't use it, and it contains UIWebView

File: Client.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -6558,7 +6471,11 @@
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
 				INFOPLIST_FILE = UnitTests/Info.plist;
-				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks @loader_path/Frameworks";
```
Comment (iccub) [inline Client.xcodeproj/project.pbxproj:6637]: XCode 11 did this automatically, apprently the format for search paths is a bit different now

File: SharedTests/UserAgentBuilderTests.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        let webView = WKWebView(frame: CGRect(x: 0, y: 0, width: 100, height: 100))
+        
+        webView.evaluateJavaScript("navigator.userAgent") { ua, error in
+            wkWebViewUA = ua as! String
+            webViewUAExpectation.fulfill()
+        }
+        
+        wait(for: [webViewUAExpectation], timeout: 10)
+        
+        XCTAssertEqual(builderUA, wkWebViewUA)
```
Comment (iccub) [inline SharedTests/UserAgentBuilderTests.swift:26]: This should give us enough confidence that our User Agent builder always match WKWebView's UA

File: Shared/UserAgentBuilder.swift [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        
+        let platform = UIDevice.current.userInterfaceIdiom == .pad ? "OS" : "iPhone OS"
+        
+        return "\(currentDevice); CPU \(platform) \(osVersion) like Mac OS X"
+    }
+    
+    var webkitVersion: UAVersions {
+        let _12_0 = OSVersionMap(majorVersion: 12, minorVersion: 0,
+                              uaVersions: UAVersions(safariVersion: "605.1",
+                                                     webkitVersion: "605.1.15"))
```
Comment (kylehickinson) [inline Shared/UserAgentBuilder.swift:54]: Are these going to be important to manually update you think?

Comment (iccub) [reply Shared/UserAgentBuilder.swift:54]: We have to.
We will see when it breaks thanks to the unit test

Chromium does it manually too I think https://chromium.googlesource.com/chromium/src.git/+/master/ios/web/common/user_agent.mm#46

**Comment by iccub** [comment]
In general I tried to utilize WKWebView first, but it's asynchronous, and can't be easily synchronized.
If Chromium uses the manual process for UA creation we should be fine imo

Old vs new UA
Zrzut ekranu 2019-08-30 o 18 06 30

**Comment by iccub** [comment]
And evaluating WKWebView's UA by javascript takes 2-3 seconds which is bad If we would like to synchronize it

**Comment by kylehickinson** [comment]
Yeah that sucks. Is there a specific spot those version numbers can be read from easily aside from watching for Chromium changes?

**Comment by iccub** [comment]
Could be done if we could subscribe to this commit history log, maybe there's github's mirror somewhere and we could subscribe to it when this file changes, no idea really

https://chromium.googlesource.com/chromium/src.git/+log/master/ios/web/common/user_agent.mm

**Comment by kylehickinson** [comment]
What if we hard-code values we know (and update based on Chromium) and overwrite with WKWebView's value async whenever it finishes successfully?

**Comment by iccub** [comment]
We can do that if we want to be extra safe. I think the unit test againt WKWebView should be enough though, and it's a bit less complex

@jhreis opinions on that welcome

**Review by kylehickinson — APPROVED** [review]
LGTM. Still wanna find a way to automate the versions somehow or at least notify us when they change rather than wait until CI tests catch it

**Review by jhreis — APPROVED** [review]

**Comment by iccub** [comment]
There's issue for up to date UA already, linking it for reference https://github.com/brave/brave-ios/issues/838

**Review by iccub — COMMENTED** [review]
I can't approve my PR, Joel's commit to add path version works fine on iOS 12.4.1

**Pull request merged**